Bug: TypeError: FakeProducer.publish() got an unexpected keyword argument 'timeout'

Hi!

**Describe the bug**

My test does not pass. I get the error: `TypeError: FakeProducer.publish() got an unexpected keyword argument 'timeout'`

Traceback:

```text

    @override
    async def publish(  # type: ignore[override]
        self,
        message: SendableMessage = "",
        reply_to: str = "",
        correlation_id: Optional[str] = None,
        headers: Optional[Dict[str, str]] = None,
        **producer_kwargs: Any,
    ) -> Optional[DecodedMessage]:
        assert self._producer, NOT_CONNECTED_YET  # nosec B101
        assert self.subject, "You have to specify outgoing subject"  # nosec B101
    
        extra: AnyDict = {
            "reply_to": reply_to or self.reply_to,
        }
        if self.stream is not None:
            extra.update(
                {
                    "stream": self.stream.name,
                    "timeout": self.timeout,
                }
            )
    
>       return await self._producer.publish(
            message=message,
            subject=self.subject,
            headers=headers or self.headers,
            correlation_id=correlation_id,
            **extra,
            **producer_kwargs,
        )
E       TypeError: FakeProducer.publish() got an unexpected keyword argument 'timeout'

```

**How to reproduce**

```python
import uuid
from datetime import datetime
from typing import Literal

from faststream.nats import NatsBroker
from pydantic import BaseModel, Field

class NatsPublisher:

    def __init__(self, settings: NATSSettings, nats_broker: NatsBroker | None = None) -> None:
        """Initialize."""
        self._nats_broker = nats_broker or NatsBroker(settings.url)

        self._publisher_gen_started = self._nats_broker.publisher(
            "generation.started", stream=settings.stream
        )
        self._publisher_gen_ended = self._nats_broker.publisher(
            "generation.ended", stream=settings.stream
        )

    async def publish_event_generation_started(
        self, payload: GenerationStartedEventModel
    ) -> None:
        """Send generation started event."""
        await self._publisher_gen_started.publish(BaseMessage(content=payload.model_dump()))

    async def publish_event_generation_ended(
        self, payload: GenerationEndedEventModel
    ) -> None:
        """Send generation ended event."""
        await self._publisher_gen_ended.publish(BaseMessage(content=payload.model_dump()))

```

Test:
```python
import pytest
from faststream.nats import NatsBroker, TestNatsBroker

@pytest.mark.asyncio()
async def test_given_nats_publisher_when_publishing_gen_start_event_then_it_works():
    # Given
    nats_settings = NATSSettings()
    broker = NatsBroker()

    @broker.subscriber("generation.started")
    def some_subscriber(msg: BaseMessage) -> None:
        _ = msg

    some_message = "some_message"

    async with TestNatsBroker(broker) as br:
        nats_publisher = NatsPublisher(nats_settings, br)

        # When
        await nats_publisher.publish_event_generation_started(some_message)  

        # Then
        assert some_subscriber.mock
        some_subscriber.mock.assert_called_once_with(some_message)
```

**Environment**
Running FastStream 0.3.8 with CPython 3.11.5 on Darwin



If the code above is not good enough to find the bug, I can write a better minimal reproducible example.

Thanks!

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -70,13 +70,14 @@ jobs:
       - name: Test
         run: bash scripts/test.sh -m "(slow and (not nats and not kafka and not rabbit and not redis)) or (not nats and not kafka and not rabbit and not redis)"
         env:
-          COVERAGE_FILE: coverage/.coverage.${{ runner.os }}-py${{ matrix.python-version }}
-          CONTEXT: ${{ runner.os }}-py${{ matrix.python-version }}
+          COVERAGE_FILE: coverage/.coverage.${{ runner.os }}-py${{ matrix.python-version }}-${{ matrix.pydantic-version }}
+          CONTEXT: ${{ runner.os }}-py${{ matrix.python-version }}-${{ matrix.pydantic-version }}
       - name: Store coverage files
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
-          name: coverage
+          name: .coverage.${{ runner.os }}-py${{ matrix.python-version }}-${{ matrix.pydantic-version }}
           path: coverage
+          if-no-files-found: error
 
   test-macos-latest:
     if: github.event.pull_request.draft == false
@@ -149,10 +150,11 @@ jobs:
           COVERAGE_FILE: coverage/.coverage.kafka-py
           CONTEXT: kafka-py
       - name: Store coverage files
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
-          name: coverage
+          name: .coverage.kafka-py
           path: coverage
+          if-no-files-found: error
 
   test-kafka-smoke:
     if: github.event.pull_request.draft == false
@@ -197,10 +199,11 @@ jobs:
           COVERAGE_FILE: coverage/.coverage.rabbit-py
           CONTEXT: rabbit-py
       - name: Store coverage files
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
-          name: coverage
+          name: .coverage.rabbit-py
           path: coverage
+          if-no-files-found: error
 
   test-rabbit-smoke:
     if: github.event.pull_request.draft == false
@@ -245,10 +248,11 @@ jobs:
           COVERAGE_FILE: coverage/.coverage.nats-py
           CONTEXT: nats-py
       - name: Store coverage files
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
-          name: coverage
+          name: .coverage.nats-py
           path: coverage
+          if-no-files-found: error
 
   test-nats-smoke:
     if: github.event.pull_request.draft == false
@@ -293,10 +297,11 @@ jobs:
           COVERAGE_FILE: coverage/.coverage.redis-py
           CONTEXT: redis-py
       - name: Store coverage files
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
-          name: coverage
+          name: .coverage.redis-py
           path: coverage
+          if-no-files-found: error
 
   test-redis-smoke:
     if: github.event.pull_request.draft == false
@@ -335,10 +340,11 @@ jobs:
           cache-dependency-path: pyproject.toml
 
       - name: Get coverage files
-        uses: actions/download-artifact@v3
+        uses: actions/download-artifact@v4
         with:
-          name: coverage
+          pattern: .coverage*
           path: coverage
+          merge-multiple: true
 
       - run: pip install coverage[toml]
 
@@ -348,7 +354,7 @@ jobs:
       - run: coverage html --show-contexts --title "FastStream coverage for ${{ github.sha }}"
 
       - name: Store coverage html
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
           name: coverage-html
           path: htmlcov

--- a/faststream/__about__.py
+++ b/faststream/__about__.py
@@ -1,5 +1,5 @@
 """Simple and fast framework to create message brokers based microservices"""
-__version__ = "0.3.8"
+__version__ = "0.3.9"
 
 
 INSTALL_YAML = """

--- a/faststream/nats/test.py
+++ b/faststream/nats/test.py
@@ -54,8 +54,10 @@ class FakeProducer(NatsFastProducer):
         subject: str,
         reply_to: str = "",
         headers: Optional[Dict[str, str]] = None,
-        stream: Optional[str] = None,
         correlation_id: Optional[str] = None,
+        # NatsJSFastProducer compatibility
+        timeout: Optional[float] = None,
+        stream: Optional[str] = None,
         *,
         rpc: bool = False,
         rpc_timeout: Optional[float] = None,
@@ -71,6 +73,9 @@ class FakeProducer(NatsFastProducer):
 
         for handler in self.broker.handlers.values():  # pragma: no branch
             call = False
+
+            if stream and getattr(handler.stream, "name", None) != stream:
+                continue
 
             if subject == handler.subject:
                 call = True

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -47,10 +47,10 @@ classifiers = [
 dynamic = ["version"]
 
 dependencies = [
-    "anyio>=3.7.1,<4; python_version < '3.11'",
-    "anyio>=3.7.1,<5; python_version >= '3.11'",
+    "anyio>=3.7.1,<5",
     "fast-depends>=2.2.6,<3",
     "typer>=0.9,<1",
+    "typing-extensions>=4.8.0",
     "uvloop>=0.18.0; sys_platform != 'win32' and (sys_platform != 'cygwin' and platform_python_implementation != 'PyPy')",
 ]
 

--- a/tests/brokers/nats/test_publish.py
+++ b/tests/brokers/nats/test_publish.py
@@ -5,4 +5,30 @@ from tests.brokers.base.publish import BrokerPublishTestcase
 
 @pytest.mark.nats
 class TestPublish(BrokerPublishTestcase):
-    pass
+    @pytest.mark.asyncio
+    async def test_stream_publish(
+        self,
+        queue: str,
+        test_broker,
+    ):
+        @test_broker.subscriber(queue, stream="test")
+        async def m():
+            ...
+
+        await test_broker.start()
+        await test_broker.publish("Hi!", queue, stream="test")
+        m.mock.assert_called_once_with("Hi!")
+
+    @pytest.mark.asyncio
+    async def test_wrong_stream_publish(
+        self,
+        queue: str,
+        test_broker,
+    ):
+        @test_broker.subscriber(queue)
+        async def m():
+            ...
+
+        await test_broker.start()
+        await test_broker.publish("Hi!", queue, stream="test")
+        assert not m.mock.called

--- a/tests/cli/test_app.py
+++ b/tests/cli/test_app.py
@@ -134,6 +134,20 @@ async def test_running(async_mock, app: FastStream):
 
 
 @pytest.mark.asyncio
+async def test_exception_group(async_mock: AsyncMock, app: FastStream):
+    app._init_async_cycle()
+    app._stop_event.set()
+
+    async_mock.excp.side_effect = ValueError("Ooops!")
+    app.on_startup(async_mock.excp)
+
+    with pytest.raises(ValueError):
+        with patch.object(app.broker, "start", async_mock.broker_run):
+            with patch.object(app.broker, "close", async_mock.broker_stopped):
+                await app.run()
+
+
+@pytest.mark.asyncio
 async def test_running_lifespan_contextmanager(async_mock, mock: Mock, app: FastStream):
     @asynccontextmanager
     async def lifespan(env: str):

--- a/tests/cli/test_app.py
+++ b/tests/cli/test_app.py
@@ -139,7 +139,10 @@ async def test_exception_group(async_mock: AsyncMock, app: FastStream):
     app._stop_event.set()
 
     async_mock.excp.side_effect = ValueError("Ooops!")
-    app.on_startup(async_mock.excp)
+
+    @app.on_startup
+    async def raises():
+        await async_mock.excp()
 
     with pytest.raises(ValueError):
         with patch.object(app.broker, "start", async_mock.broker_run):

